CCDB/include/CCDB/CCDBDownloader.h

```diff
@@ -0,0 +1,342 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <uv.h>
+#include <curl/curl.h>
+#include <string>
+#include <vector>
+#include <iostream>
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <unordered_map>
+
+#ifndef ALICEO2_CCDBDOWNLOADER_H
+#define ALICEO2_CCDBDOWNLOADER_H
+
+using namespace std;
+
+namespace o2
+{
+namespace ccdb
+{
+
+class CCDBDownloader
+{
+ public:
+
+  /**
+    * Time for which sockets will stay open after last download finishes
+    */
+  int socketTimoutMS = 4000;
+
+  /**
+   * Max number of handles that can be used at the same time
+   */
+  int maxHandlesInUse = 3;
+
+  CCDBDownloader();
+  ~CCDBDownloader();
+
+  /**
+   * Perform on a single handle in a blocking manner. Has the same effect as curl_easy_perform().
+   *
+   * @param handle Handle to be performed on. It can be reused or cleaned after perform finishes.
+   */
+  CURLcode perform(CURL* handle);
+
+  /**
+   * Perform on a batch of handles. Callback will be exectuted in it's own thread after all handles finish their transfers.
+   *
+   * @param handles Handles to be performed on.
+   */
+  std::vector<CURLcode>* asynchBatchPerformWithCallback(std::vector<CURL*> handles, bool* completionFlag, void (*cbFun)(void*), void* cbData);
+
+  /**
+   * Perform on a batch of handles in a blocking manner. Has the same effect as calling curl_easy_perform() on all handles in the vector.
+   * @param handleVector Handles to be performed on.
+   */
+  std::vector<CURLcode> batchBlockingPerform(std::vector<CURL*> handleVector);
+
+  /**
+   * Perform on a batch of handles. Completion flag will be set to true when all handles finish their transfers.
+   * @param handleVector Handles to be performed on.
+   * @param completionFlag Should be set to false before passing it to this function. Will be set to true after all transfers finish.
+   */
+  std::vector<CURLcode>* batchAsynchPerform(std::vector<CURL*> handleVector, bool* completionFlag);
+
+  /**
+   * Limits the number of parallel connections. Should be used only if no transfers are happening.
+   */
+  void setMaxParallelConnections(int limit);
+
+  /**
+   * Limits the time a socket and its connection will be opened after transfer finishes.
+   */
+  void setSocketTimoutTime(int timoutMS);
+
+ private:
+
+  /**
+   * Current amount of handles which are performed on.
+   */
+  int handlesInUse = 0;
+
+  /**
+   * Timer starts for each socket when its respective transfer finishes, and is stopped when another transfer starts for that handle.
+   * When the timer runs out it closes the socket. The period for which socket stays open is defined by socketTimoutMS.
+   */
+  std::unordered_map<curl_socket_t, uv_timer_t*> socketTimerMap;
+
+  /**
+   * The UV loop which handles transfers.
+   */
+  uv_loop_t loop;
```

Comment:
Please, make it settable from the outside or at least make sure this will not interfere with a separate libuv loop which will for sure be there.

Reply:
I have added the functionality of using an external uv_loop, which stores handles for the downloader and is cleaned of them (and only them) when the downloader is destroyed.

Could you please also elaborate what you mean by interefence between the uv_loop and the DPL one?